From 763c03163be5983fd4058b48dbd90813a9253f4e Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 23 Aug 2022 11:57:31 +0300
Subject: [PATCH] [SW-101652] gaudi3: pcie: add timeout for para sts clear

Currently, as part of SNPS DCC WA, rom loops infinitely for clear
status. As per HW team recommondation, this patch add timeout
of 10ms for PARA status to be cleared. Timeout in status clear
is recorded in ROM status update.

Change-Id: Ifd99d619c07c38c10e0c4632efadebc7cbd28d06
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index accd251..0547cb5 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -363,9 +363,10 @@
 
 #define PHY_CR_PARA_CMD_DATA		0x2
 #define PHY_CR_PARA_CMD_ADDR		0x5149
-
+#define PHY_CR_PARA_STS_UDELAY		10000 /* 10ms */
 void pcie_wa_snps_dcc(void)
 {
+	uint64_t start;
 	uint32_t val;
 
 	regmap_writel(D0_PCIE_PHY_CR_PARA_CFG_SEL_M, mmD0_PCIE_PHY_CR_PARA_CFG);
@@ -374,9 +375,19 @@
 	val = ((PHY_CR_PARA_CMD_DATA << D0_PCIE_PHY_CR_PARA_CMD_DATA_S) | PHY_CR_PARA_CMD_ADDR);
 	regmap_writel(val, mmD0_PCIE_PHY_CR_PARA_CMD);
 
+	start = arcboot_rom_get_ts_us();
 	/* Wait until status is cleared */
-	while (regmap_readl(mmD0_PCIE_PHY_CR_PARA_STS))
-		;
+	while (1) {
+		val = regmap_readl(mmD0_PCIE_PHY_CR_PARA_STS);
+		if (!val)
+			break;
+
+		if (arcboot_rom_timeout_expired(start, PHY_CR_PARA_STS_UDELAY)) {
+			arcboot_rom_status_update(
+				PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT);
+			break;
+		}
+	}
 }
 
 static void arcboot_rom_prstn_init(void)
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index aed72a9..e7159b5 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -105,40 +105,41 @@
 	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 98,
 	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 99,
 	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 100,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 101,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 102,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 103,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 104,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 105,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 106,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 107,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 108,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 109,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 110,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 111,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 112,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 113,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 114,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 115,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 116,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 117,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 118,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 119,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 120,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 121,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 122,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 123,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 124,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 125,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 126,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 127,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 128,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 129,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 130,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 131,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 132,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 133,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 134,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 135,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 136,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 101,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 102,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 103,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 104,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 105,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 106,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 107,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 108,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 109,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 110,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 111,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 112,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 113,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 114,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 115,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 116,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 117,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 118,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 119,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 120,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 121,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 122,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 123,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 124,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 125,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 126,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 127,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 128,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 129,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 130,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 131,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 132,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 133,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 134,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 135,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 136,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 137,
 };
